File: reusablebarrier.cpp
#include <stdio.h>
#include <semaphore.h>
#include <thread>
#include <vector>

using std::vector, std::thread;

sem_t barrier;
sem_t barrier2;
sem_t mutex;
int counter = 0;
int numthreads = 3;
vector<thread> threads;

void count(int thread_id) {
    
    for (int i = 0; i < numthreads; ++i) {
    
        sem_wait(&mutex);
        printf("Thread %d is at the redezvous\n", thread_id);
        counter = counter + 1;
        printf("counter before passing barrier: %d\n", counter);
        
        if (counter == numthreads) {
            sem_wait(&barrier2);
            sem_post(&barrier);
        }

        sem_post(&mutex);

        sem_wait(&barrier);
        sem_post(&barrier);

        sem_wait(&mutex);
        counter = counter - 1;
        printf("counter after passing barrier: %d\n", counter);
        printf("Thread %d is leaving barrier\n", thread_id);

        if (counter == 0) {
            sem_wait(&barrier);
            sem_post(&barrier2);
        }
        sem_post(&mutex);

        sem_wait(&barrier2);
        sem_post(&barrier2);
    }
}

int main() {
    sem_init(&barrier, 0, 0);
    sem_init(&barrier2, 0, 1);
    sem_init(&mutex, 0, 1);

    for (int i = 0; i < numthreads; ++i) {
        threads.push_back(thread(count, i));
    }

    for (thread &t : threads) {
        t.join();
    }

    sem_destroy(&barrier);
    sem_destroy(&barrier2);
    sem_destroy(&mutex);

    return 0;
}

File: reusablebarrier.py
from threading import *

barrier = Semaphore(0)
barrier2 = Semaphore(1)
mutex = Semaphore(1)
threads = []
counter = 0
numthreads = 3

def count(thread_id):
    global counter

    for i in range(numthreads):
        mutex.acquire()

        print(f"Thread {thread_id} is at the redezvous")
        counter = counter + 1
        print(f"counter before passing barrier: {counter}")
        
        if counter == numthreads:
            barrier2.acquire()
            barrier.release()
        
        mutex.release()

        barrier.acquire()
        barrier.release()

        mutex.acquire()
        
        counter = counter - 1
        print(f"counter after passing barrier: {counter}")
        print(f"Thread {thread_id} is leaving barrier")
        
        if counter == 0:
            barrier.acquire()
            barrier2.release()
        
        mutex.release()

        barrier2.acquire()
        barrier2.release()

for i in range(numthreads):
    threads.append(Thread(target=count, args=(i,)))

for thread in threads:
    thread.start()

for thread in threads:
    thread.join()

print("All threads have finished.")
